core/node/node_sync/src/external_io.rs

```diff
@@ -155,6 +155,14 @@ impl StateKeeperIO for ExternalIO {
                 )
             })?;
         let Some(mut pending_l2_block_header) = pending_l2_block_header else {
+            tracing::info!(
+                l1_batch_number = %cursor.l1_batch,
+                "Empty unsealed batch found; deleting it as we need at least one L2 block to initialize"
+            );
+            storage
+                .blocks_dal()
+                .delete_unsealed_l1_batch(cursor.l1_batch - 1)
+                .await?;
             return Ok((cursor, None));
         };
```

Comment:
An alternative here is to fetch the pending L2 block as it is guaranteed to exist on the main node (we can only receive `SyncAction::OpenBatch` command along with `SyncAction::SealL2Block`; essentially this bug only reproduces if EN restarts in between of opening a batch and processing the L2 block).

I didn't go with the approach as it seems to be error-prone and I would rather finally unblock the next release to mainnet, but happy to explore it if anyone prefers it.

Reply:
Another more robust approach is to refactor `L1BatchParamsProvider::load_l1_batch_params` in such a way that it does not rely on the first L2 block anymore as I think we don't actually need anything block-specific there. This requires much more effort and time though.

Reply:
The latter option sounds the most reasonable in the long term. Dumb question regarding alternatives: Is inserting an unsealed batch atomically with its first block not an option (in general or for ENs)? I'd imagine it would help preventing issues like this, but it could be the case that it makes an L1 batch appear in the storage too late (esp. for the main node).

Reply:
> Is inserting an unsealed batch atomically with its first block not an option (in general or for ENs)?

This was an option I considered at the start but I really wanted to minimize the amount of time when we misreport gas on low-traffic envs (the entire reason we are doing this work). We often spend more than a minute without sealing a miniblock which would mean anyone who makes gas-related requests during that time might get incorrect information.

Now, on ENs it might be less important (although above reasoning still applies to them too) but would involve some amount of refactoring on action queue logic side.

Reply:
In general I find it strange that we depend on the first miniblock so much and hopefully something I might address in the future